### PR TITLE
Lib 27 change autotask client libraries for new

### PR DESCRIPTION
**Pull request by BjMrq**
Those changes remove the authentication method used to authenticate relayers in autotask using AWS credentials.
- RelayerAutotask is removed in favor of a single Relayer client that used to be the RelayerApi client.
- The base API client can now either receive a api key and secret, in this case as before it will get the token from Cognito to authenticate requests but it can also receive an already signed token in this case it will use this pre-signed token to authenticate requests.

**IMPORTANT**
Since we remove the current way of authenticating inside autotask those changes should not be published until changes in the autotask wrapper are in place on the Defender side.
Once changes are deployed on Defender we should first publish a release candidate, and use this release candidate in autotask to make sure it works fine before publishing the changes.

### CODE REVIEW

File: packages/base/src/api/client.ts [hunk 1]
```diff
@@ -1,42 +1,68 @@
-import { AxiosInstance } from 'axios';
+import { AxiosError, AxiosInstance } from 'axios';
 import { createAuthenticatedApi } from './api';
+import { ClientCredentials, clientIsAuthenticatedWithPreSignedToken } from './auth';
+
+function sessionTokenHasExpired(httpError: AxiosError) {
+  return httpError.response?.status === 401 && httpError.response?.statusText === 'Unauthorized';
```
Comment (zeljkoX) [inline packages/base/src/api/client.ts:6]: Nice! Thanks for adding retries mechanism!

File: packages/base/src/api/client.ts [hunk 2]
```diff
@@ -1,42 +1,68 @@
...
+        apiUrl: this.getApiUrl(),
+      });
     }
     return this.api;
   }
 
-  protected async apiCall<T>(fn: (api: AxiosInstance) => Promise<T>): Promise<T> {
+  private async recreateSession() {
+    this.api = undefined;
+    return await this.init();
```
Comment (zeljkoX) [inline packages/base/src/api/client.ts:47]: We don't need to use `await` keyword here as resolution is awaited by caller.

File: packages/base/src/api/client.ts [hunk 3]
```diff
@@ -1,42 +1,68 @@
...
+  private async retryWithApiNewSession<T>(apiCallFunction: (api: AxiosInstance) => Promise<T>) {
+    const newApiSession = await this.recreateSession();
+    return await apiCallFunction(newApiSession);
+  }
+
+  protected async apiCall<T>(apiCallFunction: (api: AxiosInstance) => Promise<T>): Promise<T> {
     const api = await this.init();
     try {
-      return await fn(api);
+      return await apiCallFunction(api);
```
Comment (zeljkoX) [inline packages/base/src/api/client.ts:58]: Same here, we don't need to use await here.

File: packages/base/src/api/client.ts [hunk 4]
```diff
@@ -1,42 +1,68 @@
...
-      // this means ID token has expired so we'll recreate session and try again
-      if (error.response && error.response.status === 401 && error.response.statusText === 'Unauthorized') {
-        this.api = undefined;
-        const api = await this.init();
-        return await fn(api);
+      if (sessionTokenHasExpired(error as AxiosError)) {
+        if (this.isAuthenticatedWithPreSignedToken) {
+          throw new Error('Client expired');
+        } else {
+          return await this.retryWithApiNewSession(apiCallFunction);
```
Comment (zeljkoX) [inline packages/base/src/api/client.ts:64]: and here

File: packages/relay/src/relayer/client.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import { BaseApiClient } from 'defender-base-client';
```
Comment (zeljkoX) [inline packages/relay/src/relayer/client.ts:1]: Nice refactor!

File: packages/relay/src/autotask/index.ts [hunk 6]
```diff
@@ -1,111 +0,0 @@
-import {
```
Comment (zeljkoX) [inline packages/relay/src/autotask/index.ts:1]: Let's create a task to remove `relayer-proxy` logic from defender repo after all users have migrated to newer client versions.

**Review by zeljkoX — APPROVED** [review]
Thanks for great job!

**Comment by mok0230** [comment]
given conversation this morning, should probably close this and put it on hold for now (and look into client-side throttling)

**Comment by BjMrq** [comment]
Closed until we figure out best way to authenticate relayers in autotask

**Pull request closed without merging**